### PR TITLE
Fix StatusCancelledWithoutStartingRecvTrailingMetadata flake in client_lb_end2end_test

### PR DESCRIPTION
**Pull request by ananda1066**
Flake in `ClientLbInterceptTrailingMetadataTest.StatusCancelledWithoutStartingRecvTrailingMetadata` due to race between stream cancellation and check for trailing metadata. Added conditional var to wait on stream cancellation.

Tested with 100k runs, no flakes.

### CODE REVIEW

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 1]
```diff
@@ -2328,6 +2333,11 @@ TEST_F(ClientLbInterceptTrailingMetadataTest,
     auto stream = stub->BidiStream(&ctx);
     ctx.TryCancel();
   }
+  {
+    // Wait for stream to be cancelled.
+    grpc::internal::MutexLock lock(&mu_);
+    cond_.WaitWithTimeout(&mu_, absl::Seconds(3));
```
Comment (ctiller) [inline test/cpp/end2end/client_lb_end2end_test.cc:2339]: condition variables can wake spuriously so it's best to add a loop with a variable that becomes signalled:

```
while (!trailers_intercepted()) {
  cond_.WaitWithTimeout(&mu_, absl::Seconds(3));
}
```

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 2]
```diff
@@ -2264,10 +2267,12 @@ class ClientLbInterceptTrailingMetadataTest : public ClientLbEnd2endTest {
       self->load_report_ =
           BackendMetricDataToOrcaLoadReport(*backend_metric_data);
     }
+    self->Signal();
   }
 
+  void Signal() { cond_.Signal(); }
```
Comment (markdroth) [inline test/cpp/end2end/client_lb_end2end_test.cc:2273]: This method isn't necessary.  Instead, line 2270 above can just do `self->cond_.Signal()`.

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 3]
```diff
@@ -2251,6 +2251,9 @@ class ClientLbInterceptTrailingMetadataTest : public ClientLbEnd2endTest {
     return std::move(load_report_);
   }
 
+  grpc::internal::Mutex mu_;
```
Comment (markdroth) [inline test/cpp/end2end/client_lb_end2end_test.cc:2254]: Please make these data members private, and add a method to wait for the next event:

```
void WaitForLbCallback() {
  grpc::internal::MutexLock lock(&mu_);
  cond_.WaitWithTimeout(&mu_, absl::Seconds(3));
}
```

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 3 above)
Comment (markdroth) [reply test/cpp/end2end/client_lb_end2end_test.cc:2254]: Thinking about this further, a more general solution might be to have a queue of LB callback data and then have a method to get the latest result within some specified timeout.  For example:

```
public:
  absl::optional<LbTrailingMetadataCallbackData> GetNextTrailingMetadataCallbackData(absl::Duration timeout) {
    grpc::internal::MutexLock lock(&mu_);
    if (queue_.empty()) {
      grpc::internal::CondVar cv;
      cv_ = cv;
      if (!cv.WaitWithTimeout(&mu_, timeout)) return absl::nullopt;
    }
    auto result = std::move(queue_.front());
    queue_.pop_front();
    return result;
  }

 private:
  struct LbTrailingMetadataCallbackData { ... };

  static void ReportTrailerIntercepted(
      const grpc_core::TrailingMetadataArgsSeen& args_seen) {
    LbTrailingMetadataCallbackData entry = // ...convert from args_seen ...
    ClientLbInterceptTrailingMetadataTest* self = current_test_instance_;
    grpc::internal::MutexLock lock(&self->mu_);
    queue_.push_back(std::move(entry));
    if (cv_ != nullptr) {
      cv_->Signal();
      cv_ = nullptr;
    }
  }

  grpc::internal::Mutex mu_;
  grpc::internal::CondVar* cv_ ABSL_GUARDED_BY(&mu_) = nullptr;
  std::deque<LbTrailingMetadataCallbackData> queue_ ABSL_GUARDED_BY(&mu_);
```

This way, we're not reusing the same `CondVar` for multiple LB callbacks.

Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Hm so I have run the test many times and I'm pretty sure this test case is the only place where this race is happening - I'm wondering if we need to add this complexity? I updated with the previous review comment and the test is passing in 100k runs

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 2 above)
Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Done.

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 1 above)
Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Ah I saw this pattern in some other places and was wondering why it was needed - done

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 3 above)
Comment (markdroth) [reply test/cpp/end2end/client_lb_end2end_test.cc]: I was thinking that using the same `CondVar` for multiple LB callbacks might cause problems.  If it's not causing problems for any of the other tests, then this seems fine as-is.

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 4]
```diff
@@ -2251,24 +2251,36 @@ class ClientLbInterceptTrailingMetadataTest : public ClientLbEnd2endTest {
     return std::move(load_report_);
   }
 
+  void WaitForLbCallback() {
+    grpc::internal::MutexLock lock(&mu_);
+    while (!trailer_intercepted_) {
+      cond_.WaitWithTimeout(&mu_, absl::Seconds(3));
```
Comment (markdroth) [inline test/cpp/end2end/client_lb_end2end_test.cc:2257]: I think the `WaitForLbCallback()` method should return a bool, which should be false if `WaitWithTimeout()` returns false.  That way, the caller can say `ASSERT_TRUE(WaitForLbCallback())`, and the test will fail if no callback arrives on time.

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 3 above)
Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Yeah, this is the only test in the suite that has this issue, I haven't seen it cause any issues - will run with blase once more to make sure.

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 4 above)
Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Oh I didn't notice WaitWithTimeout() returned a bool - done, thanks!

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 3 above)
Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Ok ran with blase again - no failures for this test suite.

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 5]
```diff
@@ -2251,8 +2251,15 @@ class ClientLbInterceptTrailingMetadataTest : public ClientLbEnd2endTest {
     return std::move(load_report_);
   }
 
-  grpc::internal::Mutex mu_;
-  grpc::internal::CondVar cond_;
+  bool WaitForLbCallback() {
+    grpc::internal::MutexLock lock(&mu_);
+    bool timed_out = false;
```
Comment (markdroth) [inline test/cpp/end2end/client_lb_end2end_test.cc:2256]: I don't think we need this bool.  If the wait times out, then we know that there wasn't a callback, so there's no need to set `trailer_intercepted_` to false, which means that we can return immediately.

```
while (!trailer_intercepted_) {
  if (!cond_.WaitWithTimeout(&mu_, absl::Seconds(3))) return false;
}
trailer_intercepted_ = false;
return true;
```

**Review by markdroth — APPROVED** [review]
Just one minor comment left, otherwise looks great!

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 5 above)
Comment (ananda1066) [reply test/cpp/end2end/client_lb_end2end_test.cc]: done

**Pull request merged**